### PR TITLE
Callbacks

### PR DESCRIPTION
**Pull request by tristandeleu**
## Callbacks

The logs and history given by `keras.models.Sequential().fit()` are great ways to easily debug your models. To provide new debugging tools in the future, I am introducing callbacks for easy access to inner loops inside the training procedure.

- __Keep debug separated from the main code__: Using callbacks one can separate the training code in `fit()` from debug functions (history and logging) as they can be expressed as callbacks themselves. This improves readability and is safe for future changes.
- __Access internal states and statistics of the model__: Callbacks allow you to access internal states of the model and statistics about training *while the training procedure is still running*. You have the choice to trigger functions at any stage of training, be it at the epoch level or the batch level.

Here is an example of what you can do with callbacks, inspired by @Newmu's [tutorial](https://www.youtube.com/watch?v=S75EdAcXHKk):

cnn_activations
*Activation layers on a test example in a simple CNN during training on MNIST*

#### Why callbacks?
Callbacks are a flexible way to apply functions at any point of training. Here are some example of applications using callbacks:

- Save statistics of the training procedure, similarly to `history` in `fit()`. This callback is already available as `keras.callbacks.History`.
- Logging informations about the training procedure, depending on `verbose`. This callback is already available as `keras.callbacks.Logger` and is applied by default.
- Create animations of the weight matrices or activation layers. See below a more in-depth example.
- Get real-time informations, like the loss curve.
- Programmatically stop training ; this can be a path towards a GUI manager.

#### How to use callbacks?
To create a callback to be called during the training procedure, you only have to extend the class `keras.callbacks.Callback` with your custom functions. It supports the following methods:

- __on_train_begin__(): Method called at the beginning of training.
- __on_train_end__(): Method called at the end of training.
- __on_epoch_begin__(epoch): Method called at the beginning of epoch `epoch`.
- __on_epoch_end__(epoch, val_loss, val_acc): Method called at the end of epoch `epoch`, with validation loss `val_loss` and accuracy `val_acc` (if applicable).
- __on_batch_begin__(batch): Method called at the beginning of batch `batch`.
- __on_batch_end__(batch, indices, loss, accuracy): Method called at the end of batch `batch`, given by the indices `indices`, with loss `loss` and accuracy `accuracy` (if applicable).

In the spirit of #100 , one typical application of callbacks is to create visualizations to see how the model behaves during training. Here is an example of how you can use a callback to create an animation of a weight matrix as a function of time during training.

```python
import matplotlib.pyplot as plt
import matplotlib.animation as animation

class DrawWeights(keras.callbacks.Callback):

    def __init__(self, figsize, layer_id=0, param_id=0, weight_slice=(slice(None), 0)):
        self.layer_id = layer_id
        self.param_id = param_id
        self.weight_slice = weight_slice
        # Initialize the figure and axis
        self.fig = plt.figure(figsize=figsize)
        self.ax = self.fig.add_subplot(1, 1, 1)

    def on_train_begin(self):
        self.imgs = []

    def on_batch_end(self, batch, indices, loss, accuracy):
        # Get a snapshot of the weight matrix every 5 batches
        if batch % 5 == 0:
            # Access the full weight matrix
            weights = self.model.layers[self.layer_id].params[self.param_id].get_value()
            # Create the frame and add it to the animation
            img = self.ax.imshow(weights[self.weight_slice], interpolation='nearest')
            self.imgs.append(img)

    def on_train_end(self):
        # Once the training has ended, display the animation
        anim = animation.ArtistAnimation(self.fig, self.imgs, interval=10, blit=False)
        plt.show()
```
To apply this callback to your training procedure, simply provide the list of callbacks as an argument of `fit()`. All the callbacks given in that list are applied sequentially.
```python
draw_weights = DrawWeights(figsize=(4, 4), layer_id=0, \
    param_id=0, weight_slice=(slice(None), 0))
model.fit(X_train, Y_train, batch_size=batch_size, nb_epoch=nb_epoch, \
    verbose=1, callbacks=[draw_weights])
```
Here is a sample animation obtained with this callback:

mlp_weight
*Weight matrix of a MLP during training on MNIST*

### CODE REVIEW

**Comment by vzhong** [comment]
@tristandeleu that is super cool!

**Comment by tdhd** [comment]
Awesome indeed! :+1: 

Why not add the classes for DrawWeights and DrawActivations to the PR?

**Comment by fchollet** [comment]
This looks great! The animations look really cool. I think the callback pattern was definitely a step we had to take in order to bring about the visualization tools we're planning for. 

I will review and merge this when I'm able to (soon). For now, two questions:
- I see the fit method no longer returns the history. I think it was useful. Do you think this is incompatible with the callback pattern, or should be implemented in a different way?
- What's the performance impact of per-epoch weight matrix recovery? Theano has its own memory space, and getting objects out has a lot of overhead. So I'm curious.

File: keras/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+    def on_batch_begin(self, batch):
+        t_before_callbacks = time.time()
+        for callback in self.callbacks:
+            callback.on_batch_begin(batch)
+        self._delta_ts_batch_begin.append(time.time() - t_before_callbacks)
+        delta_t_median = np.median(self._delta_ts_batch_begin)
+        if self._delta_t_batch > 0. and delta_t_median > 0.95 * self._delta_t_batch \
+            and delta_t_median > 0.1:
+            warnings.warn('Method on_batch_begin() is slow compared '
```
Comment (fchollet) [inline keras/callbacks.py]: Pretty cool that it's checking for its own performance impact!

**Comment by fchollet** [comment]
On the API side: I think the callback system should not affect the default logging system (accessible with  the verbose flag) and the returning of the history by the `fit` method. We could re-implement these with callbacks, but I'm even sure that's necessary. Leaving the previous default logging system as is and implementing callbacks besides it should be fine (the user can setup their own logging system by setting `verbose=0` and developing a custom logger callback).

On the implementation side: for generality I think it would be better to pass values to callbacks as a dictionary where keys have explicit names, rather than as a tuple of arguments where the meaning of each value its determined by its position. For instance, this is somewhat unclear:
```python
callbacks.on_batch_end(batch_index, [], 0., 0.)
```
This would be better:
```python
callbacks.on_batch_end(batch_index, {'batch_ids':[],'loss': 0., 'acc':0.})
```
What do you think?

**Comment by tristandeleu** [comment]
Thanks everybody for the feedback!

I removed the returned history in `fit()` but to be honest I can't think of anything replacing it. So I agree with you, I should revert this change, add the `keras.callbacks.History` callback to the default callbacks in `fit()` and return the history again.  But as it is right now (with the `keras.callbacks.History` callback), the history is no longer a dict but it is an object: it changes the way to access the infos and it may break existing code using keras. Do you think it is safe to make the switch from `history` to the callback?

For logging, it is already available as a default callback `keras.callbacks.Logger` in `fit()`. There is almost no difference with the current logs as they are already expressed in some kind of "callback" way.

As for the API I fully agree with you, having a key-value input in the callback would make things easier, cleaner and safer for the future (ideally I think there should be callbacks in `evaluate()` as well). I think it makes it a better choice over predefined inputs. However it makes the choice of the key names critical that match some set of standard names and entries (`acc`/`accuracy` for example). On a similar note, the same holds with the parameters in the callbacks -- what are the parameters and how to name them.

File: keras/models.py [hunk 2]
```diff
@@ -211,67 +212,68 @@ def fit(self, X, y, batch_size=128, nb_epoch=100, verbose=1,
...
-                    history['val_acc'].append(float(val_acc))
-        return history
+                if show_accuracy:
+                    val_loss, val_acc = self.evaluate(X_val, y_val, batch_size=batch_size, \
+                        verbose=0, show_accuracy=True)
+                else:
+                    val_loss = self.evaluate(X_val, y_val, batch_size=batch_size, verbose=0)
+
+            callbacks.on_epoch_end(epoch, val_loss, val_acc)
+
```
Comment (tristandeleu) [inline keras/models.py]: Something worth noting as well is that I changed where `evaluate()` is called for validation. At the moment validation occurs at the batch level, but only for the last batch. Here validation is at the epoch level (which is equivalent to have it at the last batch), so that only `on_epoch_end()` has access to validation scores (and not `on_batch_end()`.

However it breaks the logs for validation scores since the logger update is called by `on_batch_end()` (and updating it by `on_batch_end()` and `on_epoch_end()` prints two lines somehow, I did not deeply investigate why though). So at the moment `keras.callbacks.Logger` does not display validation scores (which is the only difference with the current logging system).

**Comment by fchollet** [comment]
Merged! : ) 

I did a few changes, including:
- fixed the logging system to display validation accuracy and loss
- removed argument passing as communication mode between models and callbacks. Now callbacks can only read properties of their model, and all relevant values are stored by the model in history dictionaries. 
- Python 3 compatibility.
- bugfix (accuracy / acc)

**Comment by fchollet** [comment]
Now, onwards to create awesome visualization callbacks ;-)

**Comment by tristandeleu** [comment]
Awesome! I'm looking forward to see what people will come up with!

I see you decided to only give access to the whole history (through the model). I think callbacks should only have access to local statistics of the training, and leave it up to the callback creator to keep the history according to her needs. For example if one wants to implement a callback for an early-stopping rule (a basic rule "if the average loss on the latest epoch is <0.001 then stop learning"), it is obviously possible to retrieve the latest average loss with the whole history but that would be even easier to get the local information. In that sense the whole training history may not be a proper property of `model` since it is only used inside `fit` (besides for callbacks at the moment of course).

I also noticed that you decided to remove the exception catching in `fit`. I agree that it was out of the scope of this PR but I think it would make sense to call `on_batch_end()`, `on_epoch_end()` and `on_train_end()` if one manually stops the training. What do you think?

**Comment by fchollet** [comment]
> I see you decided to only give access to the whole history (through the model). I think callbacks should only have access to local statistics of the training, and leave it up to the callback creator to keep the history according to her needs.

I've thought about it, and I think you are right.

My reasoning for having callbacks access a global history dictionary that would be a property of the model was:
- to do other tasks, such as weight visualization, callbacks are accessing properties of the model. We might as well make this (`self.model.something`) be the unified mode of communication between callbacks and models
- we need a history object returned by the fit method
- passing a fixed number of (non-keyword) arguments to callbacks was too rigid and brittle

However, the fact that these history dictionaries would be specific to the `fit` method was an issue. We want users to be able to use a single callback while manipulating various ways of training models (eg. several `fit` in a row, or batch training using `train`). That was not possible with this approach.

So, I reverted this system, to remove these global history dictionaries and only give the callbacks access to the latest log values, through dictionaries (same system as your initial PR but with dictionaries instead of function arguments). The system is unified (every callback method takes the same argument pattern) and flexible (it's up to the callbacks to check that the parameters they need are available, and we can add any number of parameters later on with no change in API). 

As a result, the `History` callback is back. It is also being returned by default by the `fit` method.

> I also noticed that you decided to remove the exception catching in fit. I agree that it was out of the scope of this PR but I think it would make sense to call on_batch_end(), on_epoch_end() and on_train_end() if one manually stops the training. What do you think?

It does make sense. I don't have any strong opinion about this.

**Comment by lauphedo** [comment]
Amazing, but unfortunately not working on current keras version. 
Could someone please update this visualization? 
I've tried, but towards me skills are lacking...
